power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp

```diff
@@ -373,19 +390,40 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
     }
 
     // get sequence idx map of a certain batch scenario
-    SequenceIdx get_sequence_idx_map(ConstDataset const& update_data, Idx scenario_idx) const {
+    SequenceIdx get_sequence_idx_map(ConstDataset const& update_data, Idx scenario_idx,
+                                     UpdateCompIndependence comp_indenpendence = {}) const {
         auto const process_buffer_span = [](auto const& buffer_span, auto const& get_sequence) {
             auto const it_begin = buffer_span.begin();
             auto const it_end = buffer_span.end();
             return get_sequence(it_begin, it_end);
         };
 
-        auto const get_seq_idx_func = [&state = this->state_, &update_data, scenario_idx,
-                                       &process_buffer_span]<typename CT>() -> std::vector<Idx2D> {
-            auto const get_sequence = [&state](auto const& it_begin, auto const& it_end) {
-                return main_core::get_component_sequence<CT>(state, it_begin, it_end);
+        auto const get_seq_idx_func = [&state = this->state_, &update_data, scenario_idx, &process_buffer_span,
+                                       &comp_indenpendence]<typename CT>() -> std::vector<Idx2D> {
+            // TODO: (jguo) this function could be encapsulated in UpdateCompIndependence in update.hpp
+            auto const get_n_comp_elements = [&comp_indenpendence]() {
+                if (!comp_indenpendence.empty()) {
+                    auto const comp_idx = std::find_if(comp_indenpendence.begin(), comp_indenpendence.end(),
+                                                       [](auto const& comp) { return comp.name == CT::name; });
+                    if (comp_idx == comp_indenpendence.end()) {
+                        return na_Idx;
+                    }
+                    auto const& comp = *comp_idx;
+                    if (comp.is_columnar && (!comp.has_id || comp.ids_all_na)) {
+                        return comp.elements_ps_in_update;
+                    }
+                    if (!comp.is_columnar && (!comp.has_id && comp.ids_all_na)) {
+                        return comp.elements_ps_in_update;
+                    }
```

Comment:
How so? 
```
comp.is_columnar    true		false
comp.has_id false		false
comp.ids_all_na true		true
```

Reply:
I missed the `!` at the begining of the second `if` statement. My bad!

Reply:
Either way, the condition can be combined

Reply:
I would prefer if @Jerry-Jinfeng-Guo  keeps this one separated. Even though the amount of individual checks is very large (for a good reason), I feel it is a good idea to keep them that way because combining them might make it even more complex to read. Maybe a good idea could be to add a link to #548, since the checks follow the logic of the literals in 7 (here and where the errors pop).

Reply:
This conversation is closed.

Reply:
```suggestion
            // TODO: (jguo) this function could be encapsulated in UpdateCompIndependence in update.hpp
            Idx const n_comp_elements = [&comp_independence]() {
                if (!comp_independence.empty()) {
                    auto const comp_idx = std::ranges::find_if(comp_independence.begin(), comp_independence.end(),
                                                               [](auto const& comp) { return comp.name == CT::name; });
                    if (comp_idx == comp_independence.end()) {
```